<a href="https://colab.research.google.com/github/SP-TechForest/Python-creations/blob/main/NFL_DATA_PREPROCESSING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [69]:
import pandas as pd
import random 
import numpy as np 
import plotly.express as px

In [70]:
data = pd.read_csv("NFL1.csv")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning:

Columns (51) have mixed types.Specify dtype option on import or set low_memory=False.



In [71]:
data.head()

Date      GameID  Drive  qtr  ...       WPA    airWPA    yacWPA  Season
0  10-09-2009  2009091000      1    1  ...  0.060758       NaN       NaN    2009
1  10-09-2009  2009091000      1    1  ...  0.004655 -0.032244  0.036899    2009
2  10-09-2009  2009091000      1    1  ... -0.040295       NaN       NaN    2009
3  10-09-2009  2009091000      1    1  ... -0.049576  0.106663 -0.156239    2009
4  10-09-2009  2009091000      1    1  ...  0.097712       NaN       NaN    2009

[5 rows x 102 columns]

In [72]:
data.shape

(9999, 102)

Data base have 9999 rows and 102 columns 

<b>Find_missing_data</b> function will return the column name which has missing value with percentage of missing value. 

In [73]:
def find_missing_data(temp_data):
  missing_cols = data.columns[data.isna().any()]
  missing_data = data[missing_cols].isna().mean().sort_values(ascending = False)*100
  return missing_data

In [74]:
missing_data = find_missing_data(data)
Drop_cols = [missing_data.axes[0][col] for col in range(len(missing_data)) if missing_data[col] > 50 ]

Drop cols contains the column name which has missing values more than 50%. 

In [75]:
data.drop(Drop_cols, axis = 1, inplace=True)

In [76]:
data.shape

(9999, 73)

After removing the columns which have missign data more than 50%, we have 73 columns remaining. 

In [77]:
missing_data = find_missing_data(data)
missing_data

Tackler1               39.773977
down                   15.351535
DefTeamScore            7.230723
PosTeamScore            7.230723
AbsScoreDiff            7.230723
Away_WP_post            7.130713
Home_WP_post            7.130713
FirstDown               7.040704
posteam                 6.720672
DefensiveTeam           6.720672
ScoreDiff               6.720672
Win_Prob                6.710671
Away_WP_pre             6.690669
Home_WP_pre             6.690669
WPA                     1.530153
yrdline100              0.220022
yrdln                   0.220022
GoalToGo                0.220022
PlayTimeDiff            0.200020
SideofField             0.180018
EPA                     0.100010
time                    0.100010
TimeSecs                0.100010
Opp_Field_Goal_Prob     0.050005
No_Score_Prob           0.050005
Opp_Touchdown_Prob      0.050005
Field_Goal_Prob         0.050005
Safety_Prob             0.050005
Touchdown_Prob          0.050005
ExpPts                  0.050005
Opp_Safety

Table shows that, Tackler1 has 39% missign valeus. However, there columns contains name of athletes, so we cannot fill na with random values or something else. But, i am showing you the way to handle missing values in categorical columns. I am doing most frequent imputation in which i am taking most 10 frequent values and replace it with top 10 frequent values using random choice.  

In [78]:
data.Tackler1.nunique()

870

In [79]:
frequent_tackler = list(data.Tackler1.value_counts()[0:15].axes[0])

In [80]:
random.choice(frequent_tackler)

'D.Williams'

Top 10 most frequent name 

In [81]:
frequent_tackler

['D.Williams',
 'G.Hayes',
 'B.Scott',
 'L.Fletcher',
 'T.Davis',
 'J.Durant',
 'D.Jackson',
 'M.Williams',
 'C.Lofton',
 'K.Morrison',
 'T.Howard',
 'D.Manning',
 'C.Hope',
 'D.Smith',
 'Y.Bell']

In [82]:
data1 = pd.DataFrame()
data1['Tackler1'] = data['Tackler1'].fillna(random.choice(frequent_tackler)) 

In [83]:
data1.Tackler1.value_counts()[0:10]

Y.Bell        4004
D.Williams      40
G.Hayes         38
B.Scott         37
L.Fletcher      34
T.Davis         32
D.Jackson       31
J.Durant        31
M.Williams      29
C.Lofton        29
Name: Tackler1, dtype: int64

In [84]:
data.Tackler1.isna().sum()

3977

In [85]:
sample = list(np.random.choice(frequent_tackler, size = 3977, replace = True))

In [86]:
frequent_tackler = list(frequent_tackler)

In [87]:
inn = data['Tackler1'][data.Tackler1.isna()].axes[0]

most frequent values impotation 

In [88]:
for i in inn:
  data.at[i, 'Tackler1'] = np.random.choice(frequent_tackler)

In [89]:
data.Tackler1.value_counts()[0:10]

G.Hayes       313
T.Davis       311
M.Williams    311
D.Smith       309
D.Williams    308
C.Lofton      303
K.Morrison    301
L.Fletcher    294
D.Jackson     290
Y.Bell        289
Name: Tackler1, dtype: int64

In [90]:
missing_data = find_missing_data(data)

In [91]:
len(missing_data)

30

In [92]:
len(data.columns[data.isna().any()])

30

In [93]:
missing_data

down                   15.351535
AbsScoreDiff            7.230723
PosTeamScore            7.230723
DefTeamScore            7.230723
Away_WP_post            7.130713
Home_WP_post            7.130713
FirstDown               7.040704
ScoreDiff               6.720672
posteam                 6.720672
DefensiveTeam           6.720672
Win_Prob                6.710671
Away_WP_pre             6.690669
Home_WP_pre             6.690669
WPA                     1.530153
GoalToGo                0.220022
yrdline100              0.220022
yrdln                   0.220022
PlayTimeDiff            0.200020
SideofField             0.180018
time                    0.100010
EPA                     0.100010
TimeSecs                0.100010
Opp_Field_Goal_Prob     0.050005
Opp_Safety_Prob         0.050005
Opp_Touchdown_Prob      0.050005
Field_Goal_Prob         0.050005
Safety_Prob             0.050005
Touchdown_Prob          0.050005
ExpPts                  0.050005
No_Score_Prob           0.050005
dtype: flo

In [94]:
data.shape

(9999, 73)

In [95]:
data.iterrows()

<generator object DataFrame.iterrows at 0x7fd410f577d0>

In [96]:
null_rows = sum([True for inde, row in data.iterrows() if any(row.isnull())])

In [97]:
(null_rows/data.shape[0]) *100

15.791579157915791

15% rows have still missign values. <br> And, down column which has the highest missign value, has 4 unique values. 

In [98]:
data.down.nunique()

4

In [99]:
px.histogram(x = data.down)

amount of unique values in down column

In [100]:
data.down.value_counts()

1.0    3362
2.0    2520
3.0    1671
4.0     911
Name: down, dtype: int64

In [101]:
corr = data.corr()

As per the correlation data, we can say that there is no relation between down and any of the other columns of data set. 

In [102]:
corr['down']

GameID                         0.004744
Drive                          0.009354
qtr                            0.027271
down                           1.000000
TimeUnder                     -0.012793
TimeSecs                      -0.030129
PlayTimeDiff                   0.002395
yrdln                         -0.024773
yrdline100                    -0.050030
ydstogo                       -0.267057
ydsnet                        -0.007194
GoalToGo                       0.004066
FirstDown                      0.286358
PlayAttempted                       NaN
Yards.Gained                  -0.049660
sp                             0.166677
Touchdown                      0.001793
Safety                         0.012112
Onsidekick                          NaN
PassAttempt                   -0.044027
AirYards                      -0.042284
YardsAfterCatch               -0.038656
QBHit                          0.024073
InterceptionThrown             0.007671
RushAttempt                   -0.306077


In [103]:
data.drop(['down'], axis = 1, inplace = True)

In [104]:
null_rows = sum([True for inde, row in data.iterrows() if any(row.isnull())])
(null_rows/data.shape[0]) *100

13.021302130213023

After removing down column, there is 13% rows which has missign value, which i am goin to remove. 

In [105]:
null_rows

1302

In [106]:
data.dropna(inplace=True)
data.reset_index(drop=True, inplace=True)

In [107]:
find_missing_data(data)

Series([], dtype: float64)

In [108]:
data.tail()

Date      GameID  Drive  ...  Win_Prob       WPA  Season
8692  04-10-2009  2009100409      6  ...  0.835317 -0.011888    2009
8693  04-10-2009  2009100409      6  ...  0.823429  0.030023    2009
8694  04-10-2009  2009100409      6  ...  0.853452 -0.007547    2009
8695  04-10-2009  2009100409      6  ...  0.845904  0.051401    2009
8696  04-10-2009  2009100409      6  ...  0.897306 -0.008118    2009

[5 rows x 72 columns]

In [109]:
data.shape

(8697, 72)

In [110]:
100-((8697*72)/(9999*102)) * 100 

38.60327209191507

38% data loss

Initially we have 9999 rows and 102 columns, After removing null values we have 8697 rows and 72 columns. 